Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp

```diff
@@ -0,0 +1,262 @@
+    m_h_stencil_coefs_y.resize(1);
+    m_h_stencil_coefs_y[0] = 1._rt/cell_size[1];
+    m_h_stencil_coefs_z.resize(1);
+    m_h_stencil_coefs_z[0] = 1._rt/cell_size[2];
+
+    m_stencil_coefs_x.resize(m_h_stencil_coefs_x.size());
+    m_stencil_coefs_y.resize(m_h_stencil_coefs_y.size());
+    m_stencil_coefs_z.resize(m_h_stencil_coefs_z.size());
+
+    amrex::Gpu::copyAsync(amrex::Gpu::hostToDevice,
+                          m_h_stencil_coefs_x.begin(), m_h_stencil_coefs_x.end(),
+                          m_stencil_coefs_x.begin());
+    amrex::Gpu::copyAsync(amrex::Gpu::hostToDevice,
+                          m_h_stencil_coefs_y.begin(), m_h_stencil_coefs_y.end(),
+                          m_stencil_coefs_y.begin());
+    amrex::Gpu::copyAsync(amrex::Gpu::hostToDevice,
+                          m_h_stencil_coefs_z.begin(), m_h_stencil_coefs_z.end(),
+                          m_stencil_coefs_z.begin());
+    amrex::Gpu::synchronize();
+}
+
+void
+ProjectionDivCleaner::solve ()
+{
+    // Get WarpX object
+    auto & warpx = WarpX::GetInstance();
+
+    const auto& ba = warpx.boxArray();
+    const auto& dmap = warpx.DistributionMap();
+    const auto& geom = warpx.Geom();
+
+    LPInfo info;
+    info.setAgglomeration(m_agglomeration);
+    info.setConsolidation(m_consolidation);
+    info.setMaxCoarseningLevel(m_max_coarsening_level);
+
+    for (int ilev = 0; ilev < m_levels; ++ilev)
+    {
+        MLPoisson mlpoisson({geom[ilev]}, {ba[ilev]}, {dmap[ilev]}, info);
+
+        mlpoisson.setMaxOrder(m_linop_maxorder);
+
+        // This is a 3d problem with Dirichlet BC
+        mlpoisson.setDomainBC({AMREX_D_DECL(LinOpBCType::Periodic,
+                                            LinOpBCType::Periodic,
+                                            LinOpBCType::Neumann)},
+                              {AMREX_D_DECL(LinOpBCType::Periodic,
+                                            LinOpBCType::Periodic,
+                                            LinOpBCType::Neumann)});
+
+        if (ilev > 0) {
+            mlpoisson.setCoarseFineBC(m_solution[ilev-1].get(), m_ref_ratio);
+        }
+
+        m_solution[ilev]->setVal(0.);
+
+        mlpoisson.setLevelBC(ilev, m_solution[ilev].get());
+
+        MLMG mlmg(mlpoisson);
+        mlmg.setMaxIter(m_max_iter);
+        mlmg.setMaxFmgIter(m_max_fmg_iter);
+        mlmg.setVerbose(m_verbose);
+        mlmg.setBottomVerbose(m_bottom_verbose);
+        mlmg.solve({m_solution[ilev].get()}, {m_source[ilev].get()}, m_rtol, m_atol);
+
+        // Synchronize the ghost cells, do halo exchange
+        ablastr::utils::communication::FillBoundary(*(m_solution[ilev].get()),
+                                                m_solution[ilev]->nGrowVect(),
+                                                WarpX::do_single_precision_comms,
+                                                geom[ilev].periodicity());
+    }
+}
+
+void
+ProjectionDivCleaner::setSourceFromBfield ()
```

Comment:
A cleaner approach could be to add a `ComputeDivB.cpp` file (analogous to https://github.com/ECP-WarpX/WarpX/blob/development/Source/FieldSolver/FiniteDifferenceSolver/ComputeDivE.cpp) that can be called from the `run_ProjectionDivCleaner` function to populate the source. What do you think?

Reply:
I changed this to use the existing static ComputeDivB method from the WarpX class.